### ISSUE TITLE
No sound for DOS 6.22 and Windows 9x VMs

### ISSUE THREAD

**Issue by brunocastello** [utmapp/UTM#345]
I did install the required drivers for Sound Blaster 16 with DOS VM and there is no sound. There's also no sound in Windows 98SE VM, with all the devices supported (Ensoniq, SB16 and AC'97). I have drivers for all three, yet no sound is coming out after installing them. I thought it was already implemented when there was given an option to select sound device in VM settings?

Both issues can be seen after a clean install of both operating systems. I have not tried higher than Windows 9x though. Also noticed that the ReactOS VM does not have sound, too.

**Comment by osy86** [comment]
Unfortunately I get the same behavior on QEMU OSX following these instructions: https://en.wikibooks.org/wiki/QEMU/Windows_98#Virtual_machine_setup

If you can find the right args to get it working on OSX, then I can probably fix it in UTM as well.

**Comment by brunocastello** [comment]
I just did here a quick install of Windows 98 SE on QEMU. Details:

Host: 2013 rMBP 13-inch, i7, 16GB RAM, 1TB SSD, macOS Catalina latest version.
Guest: Windows 98 Second Edition (WinWorldPC ISO).

Procedure: ran on terminal the following commands:
brew install qemu
(after a while waiting for qemu to install...)
cd ~/Downloads
(it's where my ISOs are atm)
qemu-system-i386 -cdrom cdrom.iso -boot order=cd -drive file=w98.cow -m 512 -soundhw sb16 -vga cirrus -net nic,model=pcnet -net user -rtc base=localtime
(this is the command I used to run the install and after a few crashes, I managed to get to the desktop. I heard the Windows start sound and to my pleasure, I also heard the music that comes with the Windows 98 Welcome program. Actually, it was the first time I heard that music (VMware, PCem, DOSBOX-X, none of them played that music, actually).

Meaning that the SB16 drivers were actually installed and detected by Win 98 SE flawlessly.

So, all in all, the current QEMU version available on homebrew, for macOS, which is easier to use and install, can actually play music over SB16.

Details of the command I used to launch:

qemu-system-i386 
(machine i386, specifically ideal for Windows 9x, instead of x86_x64. UTM does have the i386 option, too)

-cdrom cdrom.iso 
(Windows 98 SE CD-ROM disk)

-boot order=cd 
(boot order, it's dc when you install, and cd after install)

-drive file=w98.cow 
Image of the hard disk used to install Win 98. It's a 2GB image, just for the proof of concept.

-m 512 
RAM: 512MB
-soundhw sb16 
SB16 sound card selected

-vga cirrus 
most compatible and easier to install video card on QEMU, although not very powerful (I'd rather use a S3 Trio64 or a S3 ViRGE, but both aren't available AFAIK)

-net nic,model=pcnet -net user 
(I tried to set up the network, but it actually failed for some reason. Probably because IRQ wasn't set properly - I have not checked it. I wanted to use AMD PCnet network adapter, same as used in UTM which allows me to have networking on Windows 3.11 VM)

-rtc base=localtime
Use host local time.

**Comment by osy86** [comment]
Seems like the latest QEMU is used by MacBrew (5.0.0). Not sure if that makes a difference. Any way you can test 4.2.0?

**Comment by brunocastello** [comment]
Why? When 5.0.0 works fine... it’s always better to use newer versions. Well, actually I tried to make a new win98 vm but with 8gb instead of 2gb, to accommodate some games, and all my attempts ended in a black screen before reaching the desktop, so no, I really do not want to spend another 3 hours of install attempts, trying to make a VM again for a while. This is why I like UTM, a GUI makes it easier, with a plus side - on iPad, good for travel.

The ReactOS VM you provided in another page does not have sound too. Did you have sound working there on your side?

**Comment by osy86** [comment]
UTM uses 4.2.0 so if it’s a QEMU bug that’s fixed in 5.0.0 then that would explain it.

Anyways I also can’t get Win98 to install on my Mac (stuck on windows splash after installation) so I thought you might have done something different.

ReactOS sound is not working for me on QEMU osx. Same as Windows 98.

**Comment by brunocastello** [comment]
The one successful install also was stuck on splash screen, so I just restarted it and had luck, I guess. Or maybe it had something to do with a smaller hard disk image. When I got to the desktop, sound was loud and clear. I had never heard the welcome screen app music until last night, in fact I had never heard it before - I was deaf for like 26 years until I had a cochlear implant surgery in 2008. Just recently I did a win 98 install and realised it (the welcome screen music). Sound is actually an important part of these old VMs for me because now I can get to hear all the buzz I couldn’t before when I was a kid.

**Comment by osy86** [comment]
Okay, restarted it and it did boot. But no sound here :(

**Comment by osy86** [comment]
Okay weird, I changed my audio source to my display (I had headphones) and it works now. So I guess there's some audio output source issue.

**Comment by osy86** [comment]
As expected UTM also works now (in iOS simulator). This is a QEMU/CoreAudio issue. It will have to be fixed post v1.0.

**Comment by brunocastello** [comment]
Oh, great! Thank you. I'll be waiting with a 98 VM ready to go.

**Comment by brunocastello** [comment]
While we wait for the sound fix, do you have any tips for improving Windows 98SE performance in an 2018 iPad Pro 12.9-inch? I have a game like FIFA 98 RTWC, which is a bit slow (sometimes stuttering)? I haven't tried other games yet. I was thinking if Windows 95 OSR2 would be better (less resources on background). FIFA 98 is a 3DFX/Glide game, but is playable in software renderer mode. I can't afford a 2020 iPad Pro, yet.

**Comment by brunocastello** [comment]
Meanwhile, on an unrelated topic, I remember that I had mentioned about a 3Dfx patch for QEMU in another topic. You replied to me that it was only for linux. I asked on VOGONS forums about that and this was the reply I got from them: 

> "As far as I know you would need to compile qemu for OSX on the hosts itself or another Mac. However qemu is supported on that platform. I asked the author earlier about what is needed for this patch on other platforms and opengl was the main requirement.
> I would try building from source with 3dfx patch and see if it works."

I'll test the 3Dfx patch next week when possible with QEMU, compiled on macOS, and report back in a new topic, not here. This would be a real asset in a VM, specially UTM, given that only PCem and DOSbox provide 3Dfx capabilities to games but both are way slower than QEMU. Professional virtualization solutions like VMware, Parallels and VirtualBox do not provide 3Dfx for older OSes. Basically, this is an area where UTM could have an advantage against competitors. Just by fixing the sound issue you'd have an awesome advantage over Virtualbox, for example.

That's just my 2 cents about your amazing project.

**Comment by brunocastello** [comment]
Referring from here: https://github.com/utmapp/UTM/issues/408#issuecomment-654574223

> I think sound works but you need the OS to support the QEMU sound card

Well, I have two machines - one with DOS622/WFW311 and other with Win98. Last time I tried them with latest build I had no audio (iPad Pro 2017 - previous post was a typo, not 2018 - & iPadOS 13.4.1 installed). I might try again now and play with some settings to see if I am doing something wrong. Both VMs are set to use SB16. It probably is a matter of which SB16 driver version, but I used the same I use on other virtual environments such as VMware, and macOS QEMU 5.0.0 version also worked wonders with the same drivers. Like you said previously, should be indeed a QEMU/coreaudio issue.

I think sound worked at some point in one of the previous UTM releases, but I can't remember which. One of the most recents, I believe, the RC versions. I tried reverting but no luck either.

I will update this if I find something from my new tests.

**Comment by osy86** [comment]
Sorry I was referring to 5.0.0 which I haven’t updated UTM to yet.

**Comment by brunocastello** [comment]
btw, is it safe to update iPadOS to 13.5.1 ? I did not update because I feared UTM would stop working.

**Comment by osy86** [comment]
UTM works but depending on what you’re on dolphin would not work on 13.5.1 so I wouldn’t update unless necessary.

**Comment by brunocastello** [comment]
Thanks. I have never heard of dolphin, so as long as DOS, Windows, PPC macOS and most known linux distros like Ubuntu and Debian works, I’m fine. I might just test on my iPhone first to see what happens before updating the iPad.

**Comment by brunocastello** [comment]
Just an advice, I've heard (and tested with XP SP3) that sound does work (broken) with QEMU 5.0.0 on my Mac, however the terminal keeps reporting that message "audio: Failed to create voice..." even though I can get some sound from my XP vm. I get the same with Windows 10 VM.

Further research from me found this: https://bugs.launchpad.net/qemu/+bug/1859916

Apparently with previous builds (before 5.0.0) and the next 5rc1, sound is working as intended. I will try to get the newest 5rc1 build of QEMU to try out. I think homebrew does not have it, so I will unfortunately have to find time to compile it (yuck.)

**Comment by brunocastello** [comment]
@osy86 I just came here to tell you that this issue was finally “fixed”. It was all my fault on my side. I’ll explain:

I keep my iPhone and iPad Pro on silent mode. This way they dont play the ringtones, for example the notifications or a phone call play a ringtone, these can be annoying to me sometimes, when I do not want to be disturbed. Silent mode is on at all times for that reason, giving preference to vibration mode. But I can still hear the normal sounds from videos, youtube, facebook, all other apps. For some reason, the silent mode turned on for ringtones affects UTM. Even with the volume maxed out, I had no sound. I’ve installed the latest release tonight, and still had no sound... Until I had the brilliant idea to disable the silent mode for ringtones from the control panel (a button with the symbol of a ring). As soon as I disabled it, GOTCHA!!! Windows played both start and shutdown sounds. Great! So it was probably a stupidity from my side. I’m sorry for that. But hey, deaf people problems... hehe.
